pkg/controller/atlasdeployment/search_nodes.go

```diff
@@ -0,0 +1,312 @@
+			}).Execute()
+		if err != nil {
+			return s.terminate(workflow.ErrorSearchNodesNotUpsertedInAtlas, err)
+		}
+
+		return s.progress(
+			workflow.SearchNodesUpdating,
+			fmt.Sprintf("search nodes are not ready yet: Atlas state is %q", updateResponse.GetStateName()),
+			"updating search nodes",
+		)
+	}
+	s.ctx.Log.Debug("search nodes in AKO and Atlas are equal")
+
+	// even if both Atlas and AKO are equal, check if existing atlas nodes are potentially are not IDLE (i.e. UPDATING).
+	// In this case we should continue staying in "updating" state.
+	if atlasNodes.GetStateName() != "IDLE" {
+		return s.progress(
+			workflow.SearchNodesUpdating,
+			fmt.Sprintf("search nodes are not ready yet: Atlas state is %q", atlasNodes.GetStateName()),
+			"updating search nodes",
+		)
+	}
+
+	// no changes needed, continue transitioning to pending state, marking search node status as idle.
+	return s.idle()
+}
+
+// delete deletes search nodes in Atlas. It transitions to the following states:
+// - deleting: after search nodes have been deleted in Atlas.
+// - terminated: when an error occurred.
+func (s *searchNodeController) delete() workflow.Result {
+	s.ctx.Log.Debug("deleting search nodes")
+	_, err := s.ctx.SdkClient.AtlasSearchApi.DeleteAtlasSearchDeployment(s.ctx.Context, s.projectID, s.deployment.GetDeploymentName()).Execute()
+	if err != nil {
+		return s.terminate(workflow.ErrorSearchNodesNotDeletedInAtlas, err)
+	}
+
+	return s.progress(
+		workflow.SearchNodesDeleting,
+		"deleting search nodes",
+		"deleting search nodes",
+	)
+}
+
+// progress transitions to the given state in the "SearchNodesReady" status and sets the given fineMsg as its message.
+// further it returns a coarse grained progress state for bubbling the chain.
+func (s *searchNodeController) progress(state workflow.ConditionReason, fineMsg, coarseMsg string) workflow.Result {
+	var (
+		fineProgress   = workflow.InProgress(state, fineMsg)
+		coarseProgress = workflow.InProgress(state, coarseMsg)
+	)
+
+	s.ctx.SetConditionFromResult(status.SearchNodesReadyType, fineProgress)
+	return coarseProgress
+}
+
+// terminate transitions to pending state if an error occurred.
+func (s *searchNodeController) terminate(reason workflow.ConditionReason, err error) workflow.Result {
+	s.ctx.Log.Error(err)
+	result := workflow.Terminate(reason, err.Error())
+	s.ctx.SetConditionFromResult(status.SearchNodesReadyType, result)
+	return result
+}
+
+// unmanage transitions to pending state if no search nodes are managed.
+func (s *searchNodeController) unmanage() workflow.Result {
+	s.ctx.UnsetCondition(status.SearchNodesReadyType)
+	return workflow.OK()
+}
+
+// unmanage transitions to pending state search nodes are ready and idle.
+func (s *searchNodeController) idle() workflow.Result {
+	s.ctx.SetConditionTrue(status.SearchNodesReadyType)
+	return workflow.OK()
+}
+
+func (s *searchNodeController) getAtlasSearchDeployment() (*admin.ApiSearchDeploymentResponse, bool, error) {
```

Comment:
Nit: to be a proper translation layer function this would return an internal operator struct rather than a admin.* API one. That way if the struct in the API evolves, the fix is in the translation layer.

Reply:
I am advancing what we would need in order to have a translation layer, not asking to add one right now.

But a translation layer does the following:
- Accept ONLY internal types (not API types) in function or method arguments.
- Returns ONLY internal types as results.
- Internally translate the internal argument types to API request types and the API response types back to internal types.
- Returned internal types are normalised at translation time, so they can be compared with ease.

API updates affect only code within the translation layer, unless the update is a breaking change that trickles also to internal types because some fields in use have been removed or added.

Contract tests call the translation layer and its internal types. They not need to concern with reconciliation code or the API types or functions.

Reply:
I think this is a nice proposal as part of https://docs.google.com/document/d/11yWbf1SGd5W3YeYCpDyGJ5w05nVyZhR3T_heSF2Hk_s/edit#heading=h.xwvxi8ifckvk but out of scope for this PR.